### PR TITLE
MOB-1923 Whats New Page

### PR DESCRIPTION
**Pull request by d4r1091**
[MOB-1923](https://ecosia.atlassian.net/browse/MOB-1923)

## Context
The `WhatsNewViewController` serves as an important piece of UI that informs our users about app updates and new features. It's a critical component for user engagement and education. However, the component needs to be extremely flexible in updating its items (`WhatsNewItem`), as features change or new ones get introduced. Additionally, the app has to support both iOS 13 and iOS 14+ platforms, which added another layer of complexity to the development process.

## Approach

The main goal for this update was to make the `WhatsNewViewController` highly composable and maintainable. We achieved this by:

- Introducing a ViewModel (`WhatsNewViewModel`) that feeds the ViewController with update items. This separation of concerns allows us to easily change the content without affecting the core functionality.
  
- Using AutoLayout constraints for positioning views and subviews, ensuring that the UI looks consistent across different iOS versions and device sizes.

- Making trade-offs in the design, particularly with theme properties, as our design system is evolving and the Figma components weren't up-to-date with the most recent changes. We aligned as closely as possible with the design team to make sure these trade-offs were reasonable.

- **Utilizing DataProvider**: Introduced a `DataProvider` abstraction to separate data-fetching logic from the ViewModel, enhancing testability, reusability, and future flexibility for data sources. Within the code you'll find the unused `JSONDataProvider`. Decided to leave it there to highlight the possibility of grabbing those items easily from a remote resource.

|Simulator Screenshot - iPhone 14 - 2023-09-04 at 12 59 08|Simulator Screenshot - iPhone 14 - 2023-09-04 at 12 59 05|
|:---------------:|:---------------:|

|Simulator Screenshot - iPad (10th generation) - 2023-09-04 at 15 38 00|Simulator Screenshot - iPad (10th generation) - 2023-09-04 at 15 37 51|Simulator Screenshot - iPad (10th generation) - 2023-09-04 at 15 37 54|Simulator Screenshot - iPad (10th generation) - 2023-09-04 at 15 37 58
|:---------------:|:---------------:|:---------------:|:---------------:|

### Challenges

1. **iOS Version Support**: Crafting a UI that works seamlessly across iOS 13 and iOS 14+ was a significant challenge, especially when it came to using different APIs and features available in the respective versions.

2. **Design Mismatches**: Our design system components in Figma were not perfectly up-to-date, requiring creative problem-solving and close collaboration with the design team.

3. **Theming**: Our custom theme had to be adjusted to accommodate design trade-offs. This involved deep dives into our theming logic and several iterations to get it right.

## Other

This PR is a Draft to let you see the full picture. Some components such as the `DataProvider` could be moved into Core with appropriate test scenarios written. Potentially work with Generics `T` instead of the `WhatsNewItem` solely. Small update. 

The Update logic is under evaluation.
I've followed the `DefaultBrowser` delegate mechanism and brought into this one. However, the `DefaultBrowser` was meant to be shown only within the `BrowserViewController` context. I do see this approach nestable for the WhatsNew page as well.
If we think about it, this page should be shown (as of now without any experiment nor value being provided from any remote component like Unleash etc...) within the same context as well.
Upon an app update, the iOS System kills the app. Therefore, the app will always start from the `BrowserViewController` (unless missing something). A super simple implementation could be the check whether it's been shown alongside the App Version (being at least the upcoming major 9.x.x).
Obviously the caveats here are many, as we don't want to bring the User into some unstable behaviour when multiple sheets / forms / views needs to be shown. Given a couple of tradeoffs (codebase wise) though, it seems to be achievable (already tried myself in a non-pushed codebase).
Localizations strings to be updated and added 👍 
WDYT?

### Other .1

I've noticed that the DefaultBrowser has the same issues with "trees and waves" constraints we fixed into the Share sections. WIll open a ticket and apply a simlar fix.

[MOB-1923]: https://ecosia.atlassian.net/browse/MOB-1923?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by lucaschifino — COMMENTED** [review]
Nice job, it's looking good! 👏 

Added a couple comments to the open points (data providers and app update logic), but no blockers from my side. Let me know once the PR is out of draft and ready to review so I can take a more careful look then 🙏

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        let items: [RemoteWhatsNewItem]
+    }
+
+    private struct RemoteWhatsNewItem: Decodable {
+        let imageURL: URL
+        let title: String
+        let subtitle: String
+    }
+
+    func fetchData() throws -> [WhatsNewItem] {
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:21]: Thanks for the comment @lucaschifino and the explicit no blocker for that 💪
For sure `fetchData()` might cause some misalignments in what truly does 👍 . It works mostly as an adapter component as the data will be already fetched async via another function from outside this context.+

I'm fine with removing the code, no probs at all as it will be a separate commit that could potentially be reverted.

When it comes to Braze integration we have two options as far I know/remember. 
Braze could give us a _"notification"_ telling us to do something alongside a payload of info. We might be able to receive the notification at any given/defined time. The _notification payload_ will contain all the info to show the view (which essentially will be the `Data` this provider will receive.

The other option (the one I liked the most and proposed upfront) will be to delegate to Braze the design of the entire view (this is possible via their direct SDK integration). Of course, this is now paused. Thus, the programmatically implemented WhatsNew page.

I'm inclined to leave the `DataProvider` just to better separate the concerns, as the ViewModel already contains the `items`  for a given `DataProvider` (even if just local for now).

What I'm trying to build up here, mostly, is a sort of defined, standardised and testable architecture that would make our view creation no brainer when it comes of what pattern to utilize and/or decision to make to show things like "a list".

**Comment by d4r1091** [comment]
FYI to over-clarify: the Draft is mostly to decide whether we'd like to move the `DataProvider` as part of core and make it work with Generics.

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucaschifino) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:21]: Thanks for the explanation 🙏 Braze integration delegating the design sounds indeed promising 🙌 

It's ok for me if you prefer to leave the `DataProvider` then, if you want you can maybe rethink the `fetchData` naming to avoid confusions, but it is not a must.

**Comment by lucaschifino** [comment]
> FYI to over-clarify: the Draft is mostly to decide whether we'd like to move the DataProvider as part of core and make it work with Generics.

@d4r1091 Sorry for not touching that subject earlier 🙈 

I do agree moving to Core and using generics would be a more elegant solution, but since at the moment we do not have any plans of using it anywhere else and even have the risk of not using it in the future, I see no problem in leaving it here, would even say it is better not to spend so much effort in it 🙂

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift]: Updated it into `getData`. Hope this clarifies it better its intents

**Comment by d4r1091** [comment]
The last commit https://github.com/ecosia/ios-browser/pull/546/commits/2e623ad604d878663c9123a6dfac1913b89e9a52 introduced a version management system. 
The aim was to make it easier to communicate app updates to users depending at which point they update the app from.

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    
+    /// The current version of the app.
+    private let toVersion = Version(AppInfo.ecosiaAppVersion)
+    
+    /// Default initializer.
+    init() {}
+    
+    /// The items we would like to attempt to show in the update sheet
+    private let whatsNewItems: [Version: [WhatsNewItem]] = [
+        Version("9.0.0")!: [
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:23]: Looks good with the Version link 👏

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        let items: [RemoteWhatsNewItem]
+    }
+
+    private struct RemoteWhatsNewItem: Decodable {
+        let imageURL: URL
+        let title: String
+        let subtitle: String
+    }
+
+    func getData() throws -> [WhatsNewItem] {
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:21]: I guess the version would also need to somehow be fetched and compared here to decide which items to return... Or are we assuming that whoever provides the json already has some context and filter the items out before sending?

Still, very minor since this is not currently used and only an example of how it can be in the future.

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        do {
+            let decoder = JSONDecoder()
+            decoder.keyDecodingStrategy = .convertFromSnakeCase
+            let root = try decoder.decode(WhatsNewItemRoot.self, from: jsonData)
+            return root.items.map {
+                WhatsNewItem(imageURL: $0.imageURL,
+                                        title: $0.title,
+                                        subtitle: $0.subtitle)
+            }
+        } catch {
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:33]: I guess the error here should be just thrown since this method is already throwable, shouldn't it? If that is the case there would be no reason for `do catch`.

File: Client/Ecosia/UI/WhatsNew/DataProvider/Version.swift [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    /// If no version has been saved before, it saves the `current` version passed into the function.
+    ///
+    /// - Parameters:
+    ///   - current: The current `Version` to save if no version exists. Optional.
+    ///   - prefs: The `UserDefaults` instance to use for saving and retrieving the version.
+    ///            Defaults to `UserDefaults.standard`.
+    ///
+    /// - Returns: The previously saved `Version` if it exists, otherwise returns the current version.
+    ///            Returns `nil` if both the saved version and the current version are `nil`.
+    static func retrievePreviousVersionElseSaveCurrent(_ current: Version?, using prefs: UserDefaults = UserDefaults.standard) -> Version? {
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/DataProvider/Version.swift:98]: Might have misunderstood, but isn't this always retrieving the current version instead of the previous one like the name and description suggest?
I guess the only case it return the previous is if both match 🤔 

I like the idea but found it a bit confusing to be honest... In my opinion this storing logic should not belong to `Version` itself, but to the `DataProvider`. For example `WhatsNewLocalDataProvider` would have a stored UserDefaults value like `"lastAppVersionWhereLocalWhatsNewWasShown"` and would return the items between those two versions (similar logic like you already have) and then save the next one. This is mostly due to the fact that, enabling multiple providers to consume from `Version`, we would be risking that one retrieves before the other and updates the shared stored object, which would make the second one to think the app has not updated.

I also understand I might be overcomplicating things since there should always be only one `WhatsNewDataProvider` for our view model, this is just something that comes to mind with this current concept. Just let me know if you think this is not relevant for our context.

Sorry for the long text as well, hope I managed to convey the idea and that this does not result in a long discussion topic 🙈  I'm available to chat about it as well 🙂

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        
+        // Ensure both fromVersion and toVersion are available.
+        guard let fromVersion, let toVersion else { return [] }
+        
+        // Get the version range and corresponding What's New items.
+        let versionRange = getVersionRange(from: fromVersion, to: toVersion)
+        var items: [WhatsNewItem] = []
+        for version in versionRange {
+            if let newItems = whatsNewItems[version] {
+                items.append(contentsOf: newItems)
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:50]: Nice solution already supporting aggregating versions even before it is required 🚀 ⏩

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import UIKit
+import Core
+
+final class WhatsNewCell: UITableViewCell {
+    
+    static let reuseIdentifier = String(describing: WhatsNewCell.self)
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:10]: There is a `ReusableCell` protocol that does exactly this for the cell (with the difference it instead called `cellIdentifier`). No blocker, just letting you know in case you prefer using it.

File: Client/Ecosia/UI/WhatsNew/WhatsNewViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+final class WhatsNewViewModel {
+    var items: [WhatsNewItem]
+    
+    init(provider: WhatsNewDataProvider) {
+        do {
+            self.items = try provider.getData()
+        } catch {
+            print("Failed to fetch Whats new items. Reason \(error)")
+            self.items = []
+        }
```
Comment (lucaschifino) [inline Client/Ecosia/UI/WhatsNew/WhatsNewViewModel.swift:16]: I guess with the local there is no chance of error anyway... Just a simplification suggestion, let me knwo what you think 🙂 

(Thought about adding a todo to handle error when they actually happen, but don't think it is that relevant, leaving it up to you)

```suggestion
        self.items = try? provider.getData() ?? []
```

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 9]
```diff
@@ -2766,3 +2766,10 @@ extension BrowserViewController {
                                      extras: [TelemetryWrapper.EventExtraKey.isInvertColorsEnabled.rawValue: UIAccessibility.isInvertColorsEnabled.description])
     }
 }
+
+extension BrowserViewController: WhatsNewViewDelegate {
+    
+    func whatsNewViewDidShow(_ viewController: WhatsNewViewController) {
```
Comment (lucaschifino) [inline Client/Frontend/Browser/BrowserViewController.swift:2772]: Sorry, might have missed something on the description, is this planned on being used at some point?

File: Tests/ClientTests/Ecosia/VersionTests.swift [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+@testable import Client
+
+final class VersionTests: XCTestCase {
+    
+    override class func setUp() {
+        UserDefaults.standard.removeObject(forKey: Version.appVersionUpdateKey)
+    }
+    
+    override class func tearDown() {
+        UserDefaults.standard.removeObject(forKey: Version.appVersionUpdateKey)
```
Comment (lucaschifino) [inline Tests/ClientTests/Ecosia/VersionTests.swift:15]: I guess there is no reason for `setUp` and `tearDown` to be the same? Can't you remove one of them?

**Review by lucaschifino — COMMENTED** [review]
Nice changes with the version check, really like the idea! 🚀 

Added some comments now that I went through the changes more carefully, mostly minor anyway and I guess you might check them only next week 😅

File: Client/Ecosia/L10N/String.swift [hunk 11]
```diff
@@ -230,5 +230,11 @@ extension String {
         case onboardingIllustrationTour3 = "This onboarding illustration shows a few numbers like the projects Ecosia is involved in, the total number of trees planted by the Ecosia community, alongisde the number of countries Ecosia is active. A small map of the planisphere with trees pins in few geographic location, background."
         case onboardingIllustrationTour4 = "This onboarding illustration is a photo of a monkey climbing a tree. It function mainly as general decoration image."
         case onboardingIllustrationTour4Alternative = "This onboarding illustration shows the latest financial reports of Ecosia. On the background there is an image of a person caring for tree seedlings"
+        case whatsNewViewTitle = "What's new"
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/L10N/String.swift:233]: Absolutely! That needs to be added 💯. Was planning to do it as final task.
Can add in the next commits though 👍

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift (hunk 3 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:21]: Yeah, It’s indeed only for the example proposed - will remove it as part of a separate commit so once we have any JSON support, we’ll add it back and enrich it with the version check context!

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift (hunk 4 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewJSONDataProvider.swift:33]: Resolving as irrelevant for the current context considering that the JSON Provider will be reviewed anyway in the future. 🙏

File: Client/Ecosia/UI/WhatsNew/DataProvider/Version.swift (hunk 5 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/Version.swift:98]: I’d love to catchup on that sync as I would make sure I’ve got it 💯 right! Will schedule a meeting same day as planning session to not interrupt any focus flow. Thanks for pointing out! 🙏

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift (hunk 7 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:10]: Thanks for reminder will definitely update to remove that! 🚀

File: Client/Ecosia/UI/WhatsNew/WhatsNewViewModel.swift (hunk 8 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/WhatsNewViewModel.swift:16]: Totally up -> hence committing 😝. Thanks!

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 9 above)
Comment (d4r1091) [reply Client/Frontend/Browser/BrowserViewController.swift:2772]: It’s indeed being planned to be used! I think I will likely go after a similar implementation of the current Default Browser. Bit tricky but lastly the way to go in my opinion.
Removed the `Draft` too early this time 😅

File: Tests/ClientTests/Ecosia/VersionTests.swift (hunk 10 above)
Comment (d4r1091) [reply Tests/ClientTests/Ecosia/VersionTests.swift:15]: Sure will do! Thanks for that 👀

**Comment by d4r1091** [comment]
I've been reviewing the logic to handle the `WhatsNewViewController` alongside other sheets like the `DefaultBrowserViewController`.
As mentioned, there are a few workaround and thought-thru approaches in order to handle this scenario.
Nonetheless, a bit of refactoring has been applied to make the core more readable and self-explanatory with regards to what to expect to be happening in `presentIntroViewController`.

Things I did count on were:

- it's mostly FF codebase
- we are about to the new FF vanilla
- these sheets will be likely to be removed in favor of our Engagement Services implementation that would allow us to create these views from the dashboard and decide where to appear via the SDK.
- I stopped from performing a major refactor of that as it is one of our most sensitive places (it's the browser view 😅)

File: Client/Ecosia/Extensions/BrowserViewController+Ecosia.swift [hunk 12]
```diff
@@ -36,6 +36,13 @@ extension BrowserViewController: DefaultBrowserDelegate {
     }
 }
 
+extension BrowserViewController: WhatsNewViewDelegate {
+    func whatsNewViewDidShow(_ viewController: WhatsNewViewController) {
+        profile.prefs.setInt(1, forKey: PrefsKeys.WhatsNewPageSeen)
```
Comment (ecotopian) [inline Client/Ecosia/Extensions/BrowserViewController+Ecosia.swift:41]: I would suggest to not use profile.prefs and use Core.User instead

File: Client/Ecosia/Extensions/URL+Ecosia.swift [hunk 13]
```diff
@@ -23,3 +23,39 @@ extension URL {
...
+    /// This function looks in the app's cache directory to see if an image with the given name and file extension
+    /// already exists. If it does, it returns the URL pointing to this cached image.
+    /// If the image does not exist, it attempts to create it from the app's assets and save it in the cache directory.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the image set.
+    ///   - extension: The file extension of the image (e.g., "png", "jpg", "pdf").
+    ///
+    /// - Returns: The local URL of the image in the cache directory. Returns `nil` if the operation fails.
+    static func localURLForImageset(name: String, withExtension extension: String) -> URL? {
```
Comment (ecotopian) [inline Client/Ecosia/Extensions/URL+Ecosia.swift:40]: Why do we need this? Isn't Kingfisher abstracting this already?

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        newConfiguration.textProperties.lineBreakMode = .byTruncatingTail
+        newConfiguration.secondaryText = item.subtitle
+        newConfiguration.secondaryTextProperties.lineBreakMode = .byTruncatingTail
+        newConfiguration.secondaryTextProperties.font = .preferredFont(forTextStyle: .subheadline)
+        newConfiguration.image = image
+        newConfiguration.imageProperties.maximumSize = CGSize(width: 24, height: 24)
+        contentConfiguration = newConfiguration
+    }
+    
+    private func configureForiOS13(image: UIImage?, item: WhatsNewItem) {
```
Comment (ecotopian) [inline Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:50]: Just FYI: after the next upgrade we'd need to drop support for iOS 13

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                self.configureForiOS13(image: image, item: item)
+            }
+        }
+    }
+    
+    @available(iOS 14, *)
+    private func configureForiOS14(image: UIImage?, item: WhatsNewItem) {
+        var newConfiguration = defaultContentConfiguration()
+        newConfiguration.text = item.title
+        newConfiguration.textProperties.font = .preferredFont(forTextStyle: .headline)
```
Comment (ecotopian) [inline Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:40]: I think you additionally need to specify `adjustsFontSizeToFitWidth` to make the font adjust. Also applies for secondaryTextProperties and iOS 13 UILabel.font

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 16]
```diff
@@ -2243,44 +2244,78 @@ extension BrowserViewController: UIAdaptivePresentationControllerDelegate {
...
-            // deactivate searchbar hint for new users
-            contextHintVC.viewModel.markContextualHintPresented()
+            handleFirstTimeUserActions()
         } else if User.shared.showsRebrandIntro {
-            let intro = NTPIntroViewController()
-            intro.modalPresentationStyle = .overFullScreen
-            intro.modalTransitionStyle = .crossDissolve
-            present(intro, animated: true)
-            User.shared.hideRebrandIntro()
+            presentRebrandIntro()
```
Comment (ecotopian) [inline Client/Frontend/Browser/BrowserViewController.swift:2251]: I already mentioned this in Luca's PR. We can actually remove the rebrand intro and attached logic. If it eases your task, you could include it in this task, or we'll do it in an extra one

**Review by ecotopian — COMMENTED** [review]
Great work! I skimmed through it briefly, looks great already. I added some thoughts. Will do more thorough review soon

File: Client/Ecosia/Extensions/BrowserViewController+Ecosia.swift (hunk 12 above)
Comment (d4r1091) [reply Client/Ecosia/Extensions/BrowserViewController+Ecosia.swift:41]: 💯 , as mentioned in my previous message, will definitely move it to the `Core.User` once this overall changes will be given a 👍

File: Client/Ecosia/Extensions/URL+Ecosia.swift (hunk 13 above)
Comment (d4r1091) [reply Client/Ecosia/Extensions/URL+Ecosia.swift:40]: Didn't know about Kingfisher having this functionality.
I made that so the `WhatsNewItem` will be equipped by a `imageURL: URL?` property.
Having the property as URL will provide us with a scalable model, accepting remote resources as well (in its own DataProvider).
As of now, we only have the `WhatsNewLocalDataProvider` which deals with local items. In order to be able to retrieve the local asset and evaluate the property `imageURL` correctly, I needed this extension.

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift (hunk 14 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:50]: 🎊 which is a good thing 🎉 .
This code is done so that can be easily removed though 💪

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift (hunk 15 above)
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:40]: Interesting point, double checked the NewsCell and indeed does have it. Will update it too 👍

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 16 above)
Comment (d4r1091) [reply Client/Frontend/Browser/BrowserViewController.swift:2251]: My suggestion would be to approach the removal of this Rebrand Intro logic completely as a separate tasks so that we'll have a separate PR as well

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 17]
```diff
@@ -2243,44 +2244,78 @@ extension BrowserViewController: UIAdaptivePresentationControllerDelegate {
...
-            || user.referrals.pendingClaim != nil
+                || user.referrals.pendingClaim != nil
     }
 
-    func presentDefaultBrowserPromoIfNeeded() {
-        let isHome = tabManager.selectedTab?.url.flatMap { InternalURL($0)?.isAboutHomeURL } ?? false
+    func presentInsightfulSheetsIfNeeded() {
+        if isHomePage() && presentedViewController == nil && !showLoadingScreen(for: .shared) {
+            if !presentDefaultBrowserPromoIfNeeded() {
+                presentWhatsNewPageIfNeeded()
```
Comment (lucaschifino) [inline Client/Frontend/Browser/BrowserViewController.swift:2285]: Wouldn't the whats new page have priority over the default browser promo?

Just thinking that whats new is more time pressing that default promo (which would not make a big difference to be shown next time).

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 18]
```diff
@@ -2243,44 +2244,78 @@ extension BrowserViewController: UIAdaptivePresentationControllerDelegate {
...
+        if !viewModel.items.isEmpty  {
+            WhatsNewViewController.presentOn(self)
+            return true
         }
+        return false
+    }
+
+    @discardableResult
+    private func presentDefaultBrowserPromoIfNeeded() -> Bool {
+        guard !User.shared.showsRebrandIntro, shouldShowIntroScreen, DefaultBrowserExperiment.minPromoSearches() <= User.shared.treeCount else { return false }
```
Comment (lucaschifino) [inline Client/Frontend/Browser/BrowserViewController.swift:2310]: Same as above for `User.shared.showsRebrandIntro`.

Also, got a bit confused with the `shouldShowIntroScreen` naming, somehow it does not seem to be related to default browser to me 😅 I know this already existed, just wondering if it makes sense to switch for clarity 🤔

**Review by lucaschifino — COMMENTED** [review]
Very nice changes with the version and browser logic 👏 

Added some comments but no major blocker 🚀 I can take another look once the code is final 🙂

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+import Shared
+
+/// A local data provider for fetching What's New items based on app version updates.
+final class WhatsNewLocalDataProvider: WhatsNewDataProvider {
+    
+    static let appVersionUpdateKey = "appVersionUpdateKey"
+    
+    /// The version from which the app was last updated. Optional in case this is the first run
+    /// or previous upgrading from an implmention that didn't have this one in the first place.
+    private var fromVersion: Version?
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:15]: Can do 💯 ! Thanks for the hint

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    /// - Returns: An array of `WhatsNewItem` to display.
+    func getData() throws -> [WhatsNewItem] {
+                
+        fromVersion = Version.saved(forKey: Self.appVersionUpdateKey)
+        
+        let isVersionNil = fromVersion == nil
+        let isVersionLowerThanCurrent = fromVersion != nil && fromVersion! < toVersion
+        
+        if isVersionNil ||
+            isVersionLowerThanCurrent {
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:54]: Thanks for the suggestion 🙏 . I wanted to do something similar but then realized that `Version("0.0.0")!` puts an implicit interpretation upfront.
Would rather leave it as explicit as possible.

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+            Version.updateFromCurrent(forKey: Self.appVersionUpdateKey)
+        }
+        
+        // Ensure both fromVersion is available.
+        guard let fromVersion = Version.saved(forKey: Self.appVersionUpdateKey) else { return [] }
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:59]: You are absolutely right. Thanks for spotting. I could easily do `guard let fromVersion else { return [] }` and fix it. My last update introduced a regression I didn't count on and `getData()` wasn't tested.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 17 above)
Comment (d4r1091) [reply Client/Frontend/Browser/BrowserViewController.swift:2285]: I personally agree with that 😅 - though I didn't want to modify the default behavior of the overall structure.
@ecotopian WDYT? It's an easy swap after all

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 18 above)
Comment (d4r1091) [reply Client/Frontend/Browser/BrowserViewController.swift:2310]: Wanted to keep the changes as minimal as possible but I totally agree on naming 😝

**Comment by d4r1091** [comment]
The `main` SPM commit will be reverted as the corresponding Core PR gets merged 👍

File: Client/Ecosia/Extensions/URL+Ecosia.swift (hunk 13 above)
Comment (ecotopian) [reply Client/Ecosia/Extensions/URL+Ecosia.swift:40]: Ok - I think I understand the use-case better now. Though I find the name a bit irritating for what the method does. It's state mutating and the name does not really convey this. As I see it, it serves two purpose. 
1.  Copy an asset catalogue image to the user cache (if not already there)
2. Return the cache URL

Basically you want an URL for a local image, right?

So I want to challenge this. I find it unintuitive to copy a shipped image into the user cache. Wouldn't this cause duplicate data? 

As I understand you want to leverage `URL` to have a scalable model that could either work with local or remote image urls. I see that a local URL is tricky to obtain from the Asset Catalogue. So my suggestion would be to use `data` instead of URL and make it `async`. You can retrieve `data` from `UIImage(named:)` and from a URL request. Does that make sense for your use case?

File: Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    /// from which the Ecosia App Version is retrieved
+    private let currentAppVersionProvider = DefaultAppVersionInfoProvider()
+    
+    /// Default initializer.
+    init() {}
+    
+    /// The items we would like to attempt to show in the update sheet
+    private let whatsNewItems: [Version: [WhatsNewItem]] = [
+        Version("9.0.0")!: [
+            WhatsNewItem(imageURL: URL.localURLForImageset(name: "tree", withExtension: "pdf"),
```
Comment (ecotopian) [inline Client/Ecosia/UI/WhatsNew/DataProvider/WhatsNewLocalDataProvider.swift:23]: following up on my comment from above. Instead of `imageURL` this could be `data`. It probably would make the whole getter async though.

File: Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift (hunk 14 above)
Comment (ecotopian) [reply Client/Ecosia/UI/WhatsNew/WhatsNewCell.swift:54]: exactly - very neatly separated

File: Client/Ecosia/UI/WhatsNew/WhatsNewItem.swift [hunk 23]
```diff
@@ -0,0 +1,11 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+struct WhatsNewItem {
+    let imageURL: URL?
```
Comment (ecotopian) [inline Client/Ecosia/UI/WhatsNew/WhatsNewItem.swift:8]: I guess most flexible would be `let imageData: Data?`

**Review by ecotopian — DISMISSED** [review]
Awesome work! Very well done! I do approve. There is just this one little thing.

I know that you think about extensibility already, but I have my resentments about copying asset catalogue images to the user cache. It simply does not feel like the right thing to do. So I'd would either suggest to go with a simpler approach for now (i.e. use sync data or UIImage directly in WhatsNewItem), or follow an extensible async approach (also leveraging data possibly).

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 17 above)
Comment (ecotopian) [reply Client/Frontend/Browser/BrowserViewController.swift]: I guess it's a flip coin situation. Both is important. DefaultBrowser is historically the main lever to improve and measure retention. So I'd keep this as a precedence, until we have data to prove the opposite.

**Comment by d4r1091** [comment]
Thanks for the suggestion @ecotopian !
Thanks for validating the fact that indeed I thought about scaling it as if those images would come from a remote source.
I knew the caching solution wasn't ideal when first implementing it. Theoretically I could have made a middle layer adapter that would have solved the architecture shortcut, having to deal with concrete types like `UIImage`s for the local provider solution.
It would have resulted in a much longer implementation and, considering the fact that we aim to delegate our soon-to-be-implemented Engagement Services Provider, I thought it would have been a good compromise to have a shortcut via the caching system.
The removal has been done effortlessly overall 😉 .
As I would end up to pass the `data` back and forth and transform it twice to keep the flow consistent, I opted to work with `UIImage`s directly and simplify it. 💪 
With regards to "the saving of the shown versions" solution, I opted to approach into saving the versions of the items being shown. It drastically reduced the effort of checking what to show and why.
We won't need to touch anything when we'll need to show the items and the DataProvider will take care of surfacing the items needing to be shown only. You'll find the change in the `Core` counterpart PR. 🙏

**Review by lucaschifino — COMMENTED** [review]
Looks good to me 🚀 

Just holding the approval since the core version needs to be updated after it is merged I guess 🙂

**Review by lucaschifino — APPROVED** [review]
LGTM 🚀 👏
